File: NXNavigationExtensionDemo/Feature/Advanced/Controllers/ViewController07_ScrollChangeNavigationBar.m
//
//  ViewController07_ScrollChangeNavigationBar.m
//  NXNavigationExtensionDemo
//
//  Created by Leo Lee on 2020/10/27.
//

#import <NXNavigationExtension/NXNavigationExtension.h>

#import "ViewController07_ScrollChangeNavigationBar.h"
#import "FakeNavigationBar.h"
#import "ViewController12_Modal.h"

@interface ViewController07_ScrollChangeNavigationBar () <FakeNavigationBarDelegate>

@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) FakeNavigationBar *fakeNavigationBar;

@property (nonatomic, copy) NSString *navigationBarTitle;
@property (nonatomic, assign) UIStatusBarStyle barStyle;

@property (nonatomic, strong) NSLayoutConstraint *topConstraint;
@property (nonatomic, strong) NSLayoutConstraint *leftConstraint;
@property (nonatomic, strong) NSLayoutConstraint *bottomConstraint;
@property (nonatomic, strong) NSLayoutConstraint *rightConstraint;

@end

@implementation ViewController07_ScrollChangeNavigationBar

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.tableHeaderView = self.imageView;
    
    self.navigationBarTitle = self.navigationItem.title;
    self.fakeNavigationBar.title = self.navigationItem.title;
    self.navigationItem.title = nil;
    
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(10, 40, 44, 44)];
    view.backgroundColor = [UIColor redColor];
    [self.nx_navigationBar addContainerViewSubview:self.fakeNavigationBar];
    
    self.nx_navigationBar.alpha = 0.0;
    self.barStyle = UIStatusBarStyleLightContent;
    
    [self setNeedsStatusBarAppearanceUpdate];
    
    UIView *containerView = self.nx_navigationBar.containerView;

    self.topConstraint = [self.fakeNavigationBar.topAnchor constraintEqualToAnchor:containerView.topAnchor];
    self.topConstraint.active = YES;
    
    self.leftConstraint = [self.fakeNavigationBar.leftAnchor constraintEqualToAnchor:containerView.leftAnchor];
    self.leftConstraint.active = YES;
    
    self.bottomConstraint = [self.fakeNavigationBar.bottomAnchor constraintEqualToAnchor:containerView.bottomAnchor];
    self.bottomConstraint.active = YES;
    
    self.rightConstraint = [self.fakeNavigationBar.rightAnchor constraintEqualToAnchor:containerView.rightAnchor];
    self.rightConstraint.active = YES;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    CGFloat maxHeight = CGRectGetHeight(self.nx_navigationBar.frame);
    self.tableView.contentInset = UIEdgeInsetsMake(-maxHeight, 0, 0, 0);
    
    UIEdgeInsets safeAreaInsets = self.navigationController.navigationBar.layoutMargins;
    if (@available(iOS 11.0, *)) {
        safeAreaInsets = self.navigationController.navigationBar.safeAreaInsets;
    }
    self.topConstraint.constant = safeAreaInsets.top;
    self.leftConstraint.constant = safeAreaInsets.left;
    self.bottomConstraint.constant = safeAreaInsets.bottom;
    self.rightConstraint.constant = -safeAreaInsets.right;
}

- (UIColor *)nx_barTintColor {
    return [UIColor clearColor];
}

- (BOOL)nx_containerViewWithoutNavigtionBar {
    return YES;
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return self.barStyle;
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    CGFloat offsetY = scrollView.contentOffset.y;
    if (offsetY <= 0) {
        scrollView.contentOffset = CGPointZero;
    }
    
    CGFloat imageViewHeight = CGRectGetHeight(self.imageView.frame);
    CGFloat navigationBarHeight = CGRectGetHeight(self.nx_navigationBar.frame);
    CGFloat alpha = MAX(0.0, MIN(1.0, (offsetY - imageViewHeight + navigationBarHeight) / CGRectGetHeight(self.nx_navigationBar.frame)));
    self.nx_navigationBar.alpha = alpha;
    
    if (@available(iOS 13.0, *)) {
        if (self.view.traitCollection.userInterfaceStyle == UIUserInterfaceStyleDark) {
            self.barStyle = UIStatusBarStyleLightContent;
        } else {
            self.barStyle = alpha > 0.0 ? UIStatusBarStyleDarkContent : UIStatusBarStyleLightContent;
        }
    } else {
        self.barStyle = alpha > 0.0 ? UIStatusBarStyleDefault : UIStatusBarStyleLightContent;
    }
    [self setNeedsStatusBarAppearanceUpdate];
    [self.fakeNavigationBar updateNavigationBarAlpha:alpha];
}

#pragma mark - Getter

- (UIImageView *)imageView {
    if (!_imageView) {
        _imageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"TableViewHeader"]];
        [_imageView sizeToFit];
    }
    return _imageView;
}

- (FakeNavigationBar *)fakeNavigationBar {
    if (!_fakeNavigationBar) {
        _fakeNavigationBar = [[FakeNavigationBar alloc] initWithFrame:CGRectZero];
        _fakeNavigationBar.translatesAutoresizingMaskIntoConstraints = NO;
        _fakeNavigationBar.delegate = self;
    }
    return _fakeNavigationBar;
}

#pragma mark - FakeNavigationBarDelegate

- (void)fakeNavigationBar:(FakeNavigationBar *)navigationBar didClickNavigationItemWithItemType:(FakeNavigationItemType)itemType {
    switch (itemType) {
        case FakeNavigationItemTypeBackButton: {
            [self.navigationController nx_popViewControllerAnimated:YES];
        } break;
        case FakeNavigationItemTypeAddButton: {
            UINavigationController *controller = [[[self.navigationController class] alloc] initWithRootViewController:[[ViewController12_Modal alloc] init]];
            [self presentViewController:controller animated:YES completion:NULL];
        } break;
    }
}

@end

File: Example/NXNavigationExtensionSwiftDemo/Advanced/ViewController07_ScrollChangeNavigationBar.swift
//
//  ViewController07_ScrollChangeNavigationBar.swift
//  NXNavigationExtensionSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit

class ViewController07_ScrollChangeNavigationBar: BaseTableViewController {
    
    private lazy var imageView: UIImageView = {
        let imageView = UIImageView(image: UIImage(named: "TableViewHeader"))
        imageView.sizeToFit()
        return imageView
    }()
    
    private lazy var fakeNavigationBar: FakeNavigationBar = {
        let fakeNavigationBar = FakeNavigationBar(frame: .zero)
        fakeNavigationBar.delegate = self
        fakeNavigationBar.translatesAutoresizingMaskIntoConstraints = false
        return fakeNavigationBar
    }()
    
    private var barStyle: UIStatusBarStyle = .lightContent
    private var navigationBarTitle: String?
    
    private var topConstraint: NSLayoutConstraint?
    private var leftConstraint: NSLayoutConstraint?
    private var bottomConstraint: NSLayoutConstraint?
    private var rightConstraint: NSLayoutConstraint?

    override func viewDidLoad() {
        super.viewDidLoad()

        tableView.tableHeaderView = imageView
        
        navigationBarTitle = navigationItem.title
        fakeNavigationBar.title = navigationItem.title
        navigationItem.title = nil
        
        let view = UIView(frame: CGRect(x: 10, y: 40, width: 44, height: 44))
        view.backgroundColor = .red
        nx_navigationBar.addContainerSubview(fakeNavigationBar)
        
        nx_navigationBar.alpha = 0.0
        barStyle = .lightContent
        
        setNeedsStatusBarAppearanceUpdate()

        let containerView = nx_navigationBar.containerView
        
        topConstraint = fakeNavigationBar.topAnchor.constraint(equalTo: containerView.topAnchor)
        topConstraint?.isActive = true
        
        leftConstraint = fakeNavigationBar.leftAnchor.constraint(equalTo: containerView.leftAnchor)
        leftConstraint?.isActive = true
        
        bottomConstraint = fakeNavigationBar.bottomAnchor.constraint(equalTo: containerView.bottomAnchor)
        bottomConstraint?.isActive = true
        
        rightConstraint = fakeNavigationBar.rightAnchor.constraint(equalTo: containerView.rightAnchor)
        rightConstraint?.isActive = true
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        
        let maxHeight = nx_navigationBar.frame.height
        tableView.contentInset = UIEdgeInsets(top: -maxHeight, left: 0.0, bottom: 0.0, right: 0.0)
        
        let safeAreaInsets = navigationController?.navigationBar.safeAreaInsets ?? .zero
        topConstraint?.constant = safeAreaInsets.top
        leftConstraint?.constant = safeAreaInsets.left
        bottomConstraint?.constant = safeAreaInsets.bottom
        rightConstraint?.constant = -safeAreaInsets.right
    }
    
    override var nx_barTintColor: UIColor? {
        return .clear
    }
    
    override var nx_containerViewWithoutNavigtionBar: Bool {
        return true
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return barStyle
    }
    
}

extension ViewController07_ScrollChangeNavigationBar {
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let offsetY = scrollView.contentOffset.y
        if offsetY <= 0 {
            scrollView.contentOffset = .zero
        }

        let imageViewHeight = imageView.frame.height
        let navigationBarHeight = nx_navigationBar.frame.height
        let alpha = CGFloat.maximum(0.0, CGFloat.minimum(1.0, (offsetY - imageViewHeight + navigationBarHeight) / nx_navigationBar.frame.height))

        nx_navigationBar.alpha = alpha

        if #available(iOS 13.0, *) {
            if view.traitCollection.userInterfaceStyle == .dark {
                barStyle = .lightContent
            } else {
                barStyle = alpha > 0.0 ? .darkContent : .lightContent
            }
        } else {
            barStyle = alpha > 0.0 ? .default : .lightContent
        }

        setNeedsStatusBarAppearanceUpdate()
        fakeNavigationBar.updateNavigationBar(alpha: alpha)
    }
    
}

extension ViewController07_ScrollChangeNavigationBar: FakeNavigationBarDelegate {
    
    func fakeNavigationBar(_ navigationBar: FakeNavigationBar, didClickNavigationItemWithItemType itemType: FakeNavigationBar.ItemType) {
        switch itemType {
        case .back:
            navigationController?.nx_triggerSystemBackButtonHandler()
        case .add:
            guard let navigationController = navigationController else { return }
            let controller = type(of: navigationController).init(rootViewController: ViewController12_Modal())
            present(controller, animated: true, completion: nil)
        }
    }
    
}
